engine/src/rocksdb_engine.rs

```diff
@@ -0,0 +1,197 @@
+        db_opts.create_if_missing(true);
+        Ok(Self {
+            inner: Arc::new(
+                DB::open_cf(&db_opts, data_dir, tables).map_err(|e| {
+                    EngineError::UnderlyingError(format!("cannot open database: {e}"))
+                })?,
+            ),
+        })
+    }
+}
+
+impl StorageEngine for RocksEngine {
+    #[inline]
+    fn get(&self, table: &str, key: impl AsRef<[u8]>) -> Result<Option<Vec<u8>>, EngineError> {
+        if let Some(cf) = self.inner.cf_handle(table) {
+            Ok(self
+                .inner
+                .get_cf(&cf, key)
+                .map_err(|e| EngineError::IoError(format!("get key from {table} failed: {e}")))?)
+        } else {
+            Err(EngineError::TableNotFound(table.to_owned()))
+        }
+    }
+
+    #[inline]
+    fn get_multi(
+        &self,
+        table: &str,
+        keys: &[impl AsRef<[u8]>],
+    ) -> Result<Vec<Option<Vec<u8>>>, EngineError> {
+        if let Some(cf) = self.inner.cf_handle(table) {
+            self.inner
+                .multi_get_cf(repeat(&cf).zip(keys.iter()))
+                .into_iter()
+                .map(|res| {
+                    res.map_err(|err| {
+                        EngineError::IoError(format!("get key from {table} failed: {err}"))
+                    })
+                })
+                .collect::<Result<Vec<_>, _>>()
+        } else {
+            Err(EngineError::TableNotFound(table.to_owned()))
+        }
+    }
+
+    #[inline]
+    fn write_batch(&self, wr_ops: Vec<WriteOperation<'_>>, sync: bool) -> Result<(), EngineError> {
+        let mut batch = WriteBatchWithTransaction::<false>::default();
+
+        for op in wr_ops {
+            match op {
+                WriteOperation::Put(Put { table, key, value }) => {
+                    let cf = self
+                        .inner
+                        .cf_handle(table)
+                        .ok_or(EngineError::TableNotFound(table.to_owned()))?;
+                    batch.put_cf(&cf, key, value);
+                }
+                WriteOperation::Delete(Delete { table, key }) => {
+                    let cf = self
+                        .inner
+                        .cf_handle(table)
+                        .ok_or(EngineError::TableNotFound(table.to_owned()))?;
+                    batch.delete_cf(&cf, key);
+                }
+                WriteOperation::DeleteRange(DeleteRange { table, from, to }) => {
+                    let cf = self
+                        .inner
+                        .cf_handle(table)
+                        .ok_or(EngineError::TableNotFound(table.to_owned()))?;
+                    batch.delete_range_cf(&cf, from, to);
+                }
+            }
+        }
+        let mut opt = WriteOptions::default();
+        opt.set_sync(sync);
+        self.inner
+            .write_opt(batch, &opt)
+            .map_err(|e| EngineError::UnderlyingError(format!("{e}")))
```

Comment:
What's the difference between `UnderlyingError` and `IoError`? I notice we will return `IoError` in the above code.

Reply:
Rocksdb provides 16 different errors. The `UnderlyingError` is a storage engine-specific error wrapper.